### PR TITLE
User step

### PR DESCRIPTION
**Pull request by FlanFlanagan**
Adds the user 'decision step' to the time_listener. 

Addresses modification to CEP20

### CODE REVIEW

**Comment by FlanFlanagan** [comment]
I have an update for this tomorrow. (Please don't merge yet).

File: src/time_listener.h [hunk 1]
```diff
@@ -33,6 +33,12 @@ class TimeListener: virtual public Ider {
   ///
   /// @param time is the current simulation timestep
   virtual void Tock() = 0;
+
+  /// Simulation agents do their end-of-timestep decisions in the
```
Comment (katyhuff) [inline src/time_listener.h]: Perhaps some more info on what constitutes a 'decision' might be appropriate for this docstring.

File: tests/test_bear_deploy.py [hunk 2]
```diff
@@ -19,8 +19,8 @@
    'config': {
     'DemandFac': {
      'commodity': 'bears',
-     'production_rate_max': 12,
-     'production_rate_min': 8,
+     'production_rate_max': 10,
+     'production_rate_min': 10,
```
Comment (katyhuff) [inline tests/test_bear_deploy.py]: same comment here as in #1477

File: tests/test_bear_deploy.py [hunk 3]
```diff
@@ -31,7 +31,7 @@
     'config': {
      'NOInst': {
       'growth_commod': 'bears',
-      'growth_rate': 0.05,
+      'growth_rate': 0.1,
```
Comment (katyhuff) [inline tests/test_bear_deploy.py]: why change this test in this way? (Sorry to be picky, but changes in the tests are usually a little bit of a no-no for me.) Adding tests is always fine, but altering or removing tests can allow changes to introduce backwards incompatibility, loss of functionality, etc.

File: tests/test_bear_deploy.py [hunk 4]
```diff
@@ -57,7 +57,7 @@ def test_bear_deploy():
     # test that the institution deploys a BearStore
     assert_in("New fac: BearStore", s)
     # test that the first agents exist with right minimum production.
-    agents = re.compile('Agent \d+ 8\.0')
+    agents = re.compile('Agent \d+ 10\.0')
```
Comment (katyhuff) [inline tests/test_bear_deploy.py]: why change this test in this way? (Sorry to be picky, but changes in the tests are usually a little bit of a no-no for me.) Adding tests is always fine, but altering or removing tests can allow changes to introduce backwards incompatibility, loss of functionality, etc.

**Comment by gwenchee** [comment]
Before merging, a test for the decision phase should be added?

**Comment by jbae11** [comment]
@gwenchee and I tested this Cyclus version on different inputs, and only the ones without any material exchange succeeds. (like [this file](https://github.com/ergs/d3ploy/blob/develop/tests/test.xml)). 

All simulations with material exchange returns a Segmentation fault on the decision phase.

I made a small working example with just a cycamore source and sink: 
[gist](https://gist.github.com/jbae11/ad3607a7ec7325dd384a81d6c0dfb748)
the input file and the error message is in the gist.

**Comment by FlanFlanagan** [comment]
I'm not seeing this issue at all. Can you tell me which branches of everything you're on?

**Comment by FlanFlanagan** [comment]
Also, can you run cyclus in gdb, that should tell use where the segfault is occurring.  Typically a segmentation fault all by itself is pretty difficult to diagnose.

**Comment by gwenchee** [comment]
Flanflanagan/cyclus --> user-step branch 
Flanflanagan/cycamore --> d3ploy-update
ergs/d3ploy --> develop

**Comment by FlanFlanagan** [comment]
Try, d3ploy flanflanagan/decision, and d3ploy update was merged into cycamore master.

**Comment by gwenchee** [comment]
Oh yes i meant 
flanflanagan/d3ploy --> decision

**Comment by FlanFlanagan** [comment]
Okay, well then I need the gdb readout then.

**Comment by jbae11** [comment]
how do we run that in cyclus?

**Comment by FlanFlanagan** [comment]
gdb cyclus
run <inputfile>

**Comment by jbae11** [comment]
@FlanFlanagan I'm not sure I did it right but this is what I got:
```
teddy ~/Desktop $ gdb cyclus
GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from cyclus...(no debugging symbols found)...done.
(gdb) run z.xml
Starting program: /home/teddy/.local/bin/cyclus z.xml
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffe7fa5700 (LWP 4747)]
[New Thread 0x7fffe77a4700 (LWP 4748)]
[New Thread 0x7fffe4fa3700 (LWP 4749)]
[Thread 0x7fffe4fa3700 (LWP 4749) exited]
[Thread 0x7fffe77a4700 (LWP 4748) exited]
[Thread 0x7fffe7fa5700 (LWP 4747) exited]
              :                                                               
          .CL:CC CC             _Q     _Q  _Q_Q    _Q    _Q              _Q   
        CC;CCCCCCCC:C;         /_\)   /_\)/_/\\)  /_\)  /_\)            /_\)  
        CCCCCCCCCCCCCl       __O|/O___O|/O_OO|/O__O|/O__O|/O____________O|/O__
     CCCCCCf     iCCCLCC     /////////////////////////////////////////////////
     iCCCt  ;;;;;.  CCCC                                                      
    CCCC  ;;;;;;;;;. CClL.                          c                         
   CCCC ,;;       ;;: CCCC  ;                   : CCCCi                       
    CCC ;;         ;;  CC   ;;:                CCC`   `C;                     
  lCCC ;;              CCCC  ;;;:             :CC .;;. C;   ;    :   ;  :;;   
  CCCC ;.              CCCC    ;;;,           CC ;    ; Ci  ;    :   ;  :  ;  
   iCC :;               CC       ;;;,        ;C ;       CC  ;    :   ; .      
  CCCi ;;               CCC        ;;;.      .C ;       tf  ;    :   ;  ;.    
  CCC  ;;               CCC          ;;;;;;; fC :       lC  ;    :   ;    ;:  
   iCf ;;               CC         :;;:      tC ;       CC  ;    :   ;     ;  
  fCCC :;              LCCf      ;;;:         LC :.  ,: C   ;    ;   ; ;   ;  
  CCCC  ;;             CCCC    ;;;:           CCi `;;` CC.  ;;;; :;.;.  ; ,;  
    CCl ;;             CC    ;;;;              CCC    CCL                     
   tCCC  ;;        ;; CCCL  ;;;                  tCCCCC.                      
    CCCC  ;;     :;; CCCCf  ;                     ,L                          
     lCCC   ;;;;;;  CCCL                                                      
     CCCCCC  :;;  fCCCCC                                                      
      . CCCC     CCCC .                                                       
       .CCCCCCCCCCCCCi                                                        
          iCCCCCLCf                                                           
           .  C. ,                                                            
              :                                                               

Thread 1 "cyclus" received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) cyclus z.xml
Undefined command: "cyclus".  Try "help".
(gdb) run z.xml
The program being debugged has been started already.
Start it from the beginning? (y or n) y
Starting program: /home/teddy/.local/bin/cyclus z.xml
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffe7fa5700 (LWP 4753)]
[New Thread 0x7fffe77a4700 (LWP 4754)]
[New Thread 0x7fffe2fa3700 (LWP 4755)]
[Thread 0x7fffe2fa3700 (LWP 4755) exited]
[Thread 0x7fffe77a4700 (LWP 4754) exited]
[Thread 0x7fffe7fa5700 (LWP 4753) exited]
              :                                                               
          .CL:CC CC             _Q     _Q  _Q_Q    _Q    _Q              _Q   
        CC;CCCCCCCC:C;         /_\)   /_\)/_/\\)  /_\)  /_\)            /_\)  
        CCCCCCCCCCCCCl       __O|/O___O|/O_OO|/O__O|/O__O|/O____________O|/O__
     CCCCCCf     iCCCLCC     /////////////////////////////////////////////////
     iCCCt  ;;;;;.  CCCC                                                      
    CCCC  ;;;;;;;;;. CClL.                          c                         
   CCCC ,;;       ;;: CCCC  ;                   : CCCCi                       
    CCC ;;         ;;  CC   ;;:                CCC`   `C;                     
  lCCC ;;              CCCC  ;;;:             :CC .;;. C;   ;    :   ;  :;;   
  CCCC ;.              CCCC    ;;;,           CC ;    ; Ci  ;    :   ;  :  ;  
   iCC :;               CC       ;;;,        ;C ;       CC  ;    :   ; .      
  CCCi ;;               CCC        ;;;.      .C ;       tf  ;    :   ;  ;.    
  CCC  ;;               CCC          ;;;;;;; fC :       lC  ;    :   ;    ;:  
   iCf ;;               CC         :;;:      tC ;       CC  ;    :   ;     ;  
  fCCC :;              LCCf      ;;;:         LC :.  ,: C   ;    ;   ; ;   ;  
  CCCC  ;;             CCCC    ;;;:           CCi `;;` CC.  ;;;; :;.;.  ; ,;  
    CCl ;;             CC    ;;;;              CCC    CCL                     
   tCCC  ;;        ;; CCCL  ;;;                  tCCCCC.                      
    CCCC  ;;     :;; CCCCf  ;                     ,L                          
     lCCC   ;;;;;;  CCCL                                                      
     CCCCCC  :;;  fCCCCC                                                      
      . CCCC     CCCC .                                                       
       .CCCCCCCCCCCCCi                                                        
          iCCCCCLCf                                                           
           .  C. ,                                                            
              :                                                               

Thread 1 "cyclus" received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) 
```

**Comment by FlanFlanagan** [comment]
Sorry, once you get the error you need to type 'bt'

This will return a backtrace of where the error occured.

**Comment by jbae11** [comment]
thanks
```
teddy ~/Desktop $ gdb cyclus
GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from cyclus...(no debugging symbols found)...done.
(gdb) run z.xml
Starting program: /home/teddy/.local/bin/cyclus z.xml
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffe7fa5700 (LWP 5416)]
[New Thread 0x7fffe77a4700 (LWP 5417)]
[New Thread 0x7fffe4fa3700 (LWP 5418)]
[Thread 0x7fffe4fa3700 (LWP 5418) exited]
[Thread 0x7fffe77a4700 (LWP 5417) exited]
[Thread 0x7fffe7fa5700 (LWP 5416) exited]
              :                                                               
          .CL:CC CC             _Q     _Q  _Q_Q    _Q    _Q              _Q   
        CC;CCCCCCCC:C;         /_\)   /_\)/_/\\)  /_\)  /_\)            /_\)  
        CCCCCCCCCCCCCl       __O|/O___O|/O_OO|/O__O|/O__O|/O____________O|/O__
     CCCCCCf     iCCCLCC     /////////////////////////////////////////////////
     iCCCt  ;;;;;.  CCCC                                                      
    CCCC  ;;;;;;;;;. CClL.                          c                         
   CCCC ,;;       ;;: CCCC  ;                   : CCCCi                       
    CCC ;;         ;;  CC   ;;:                CCC`   `C;                     
  lCCC ;;              CCCC  ;;;:             :CC .;;. C;   ;    :   ;  :;;   
  CCCC ;.              CCCC    ;;;,           CC ;    ; Ci  ;    :   ;  :  ;  
   iCC :;               CC       ;;;,        ;C ;       CC  ;    :   ; .      
  CCCi ;;               CCC        ;;;.      .C ;       tf  ;    :   ;  ;.    
  CCC  ;;               CCC          ;;;;;;; fC :       lC  ;    :   ;    ;:  
   iCf ;;               CC         :;;:      tC ;       CC  ;    :   ;     ;  
  fCCC :;              LCCf      ;;;:         LC :.  ,: C   ;    ;   ; ;   ;  
  CCCC  ;;             CCCC    ;;;:           CCi `;;` CC.  ;;;; :;.;.  ; ,;  
    CCl ;;             CC    ;;;;              CCC    CCL                     
   tCCC  ;;        ;; CCCL  ;;;                  tCCCCC.                      
    CCCC  ;;     :;; CCCCf  ;                     ,L                          
     lCCC   ;;;;;;  CCCL                                                      
     CCCCCC  :;;  fCCCCC                                                      
      . CCCC     CCCC .                                                       
       .CCCCCCCCCCCCCi                                                        
          iCCCCCLCf                                                           
           .  C. ,                                                            
              :                                                               

Thread 1 "cyclus" received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00007fffeceb91bb in cyclus::Timer::DoTick() () from /home/teddy/.local/lib/libcyclus.so
#2  0x00007fffeceb87ce in cyclus::Timer::RunSim() () from /home/teddy/.local/lib/libcyclus.so
#3  0x00000000004385ac in main ()
(gdb) 
```

**Comment by FlanFlanagan** [comment]
So the issue is occurring in the Tick() phase. So there might be something in the tick of Cycamore:Source or Cycamore:Sink. What's odd is that I'm not seeing this. My current configuration is as follows. 

cyclus - > flanflanagan/userstep
cycamore - > cyclus / master
d3ploy - > flanflanagan / decision

**Comment by jbae11** [comment]
yup. I have the same configuration.

**Comment by gwenchee** [comment]
I just tried again and I think i had a bad cycamore branch when i first did it. It works for me now! Sorry for the confusion @FlanFlanagan

**Comment by jbae11** [comment]
yup. runs for me as well.

**Comment by FlanFlanagan** [comment]
Well then, can we get this merged in?

**Comment by gwenchee** [comment]
Yes!

**Pull request merged**